KeyboardShortcuts: Update delete shortcut to use `shift + Backspace`

yogeshbhutkar opened the pull request:
Fixes: #68139 

## What, Why & How?
Previously, the shortcut to remove the `selected blocks` was `Control` `Option` `Z` on Mac, or `Control` `Alt` `Z` on Windows.

This PR introduces the usage of `shift` + `Backspace` for performing the same.

## Testing Instructions
1. Create a block.
2. Press the key combination `SHIFT` + `Backspace` and observe the removal of the block.

## Screencast


https://github.com/user-attachments/assets/0600431d-8b16-44f1-abf0-ed58a455617a





## Screenshot

![Screenshot 2024-12-24 at 3 15 55 PM](https://github.com/user-attachments/assets/da293c9a-5ad7-4d9f-ad09-ce5e1cb7fd93)

yogeshbhutkar committed:
KeyboardShortcuts: Update delete shortcut to use Shift + Delete and add Backspace alias

yogeshbhutkar committed:
KeyboardShortcuts: Add Shift modifier for Backspace alias

yogeshbhutkar committed:
KeyboardShortcuts: Update delete shortcut character from 'del' to 'delete'

yogeshbhutkar commented:
Looks like some Playwright tests are failing, will be adding a commit to fix the failing test cases.

yogeshbhutkar committed:
KeyboardShortcuts: Update delete shortcut to use Shift + Delete

yogeshbhutkar committed:
KeyboardShortcuts: Change delete shortcut to platformAlt + Backspace

yogeshbhutkar committed:
KeyboardShortcuts: Update delete shortcut to use primaryShift + Delete

yogeshbhutkar committed:
KeyboardShortcuts: Update Block Locking spec to use the proper key shortcut

jasmussen commented:
Took the latest iteration for a spin. This feel very solid for me:

https://github.com/user-attachments/assets/4f8ec017-8c24-4135-bc14-3b49c949a8fd

I'd encourage others to give this a spin, to check compatibility across platforms, etc, but this PR feels more intuitive than trunk, for the moment. Thank you!

t-hamano commented:
On Windows, the shortcut looks like this:

![image](https://github.com/user-attachments/assets/b8b5b60a-a1a4-453f-9438-aa679225039a)

One thing I noticed is that depending on the cursor position, the process to merge blocks can be prioritized. Below are two scenarios:

- Execute <kbd>ctrl</kbd>+<kbd>shift</kbd>+<kbd>backspace</kbd>at the beginning of the text
- Execute <kbd>ctrl</kbd>+<kbd>shift</kbd>+<kbd>del</kbd> at the end of the text


https://github.com/user-attachments/assets/c2ffadc1-0307-4ebe-a124-dc2baf50a415


Can these be reproduced on a Mac? If the <kbd>backspace</kbd> and <kbd>del</kbd> keys don't work consistently, we might want to consider the <kbd>d</kbd> key.

yogeshbhutkar commented:
> Can these be reproduced on a Mac? 

Yes, @t-hamano, you are right about that. And it's the same for Mac. I believed it was intended and that `backspace` and `del` should be aliased together.

> we might want to consider the d key.

`ctrl+alt+d` and `ctrl+shift+alt+d` seem to be the only shortcuts available with `d`. The second one is too big, leaving just `ctrl+alt+d` behind.

What are your thoughts on this? Do we proceed with implementing `ctrl+alt+d`?

t-hamano commented:
I came up with a different solution: merging blocks using <kbd>del</kbd> or <kbd>Backspace</kbd> is done [here](https://github.com/WordPress/gutenberg/blob/c3ca59b70c9807e78b45d22d0714be777da507fd/packages/block-editor/src/components/rich-text/event-listeners/delete.js#L17), so when <kbd>command/ctrl</kbd> + <kbd>shift</kbd> + <kbd>del/backspace</kbd> is executed, aborts the process:

```diff
diff --git a/packages/block-editor/src/components/rich-text/event-listeners/delete.js b/packages/block-editor/src/components/rich-text/event-listeners/delete.js
index ae3fd733bb..12512626d9 100644
--- a/packages/block-editor/src/components/rich-text/event-listeners/delete.js
+++ b/packages/block-editor/src/components/rich-text/event-listeners/delete.js
@@ -6,7 +6,7 @@ import { isCollapsed, isEmpty } from '@wordpress/rich-text';
 
 export default ( props ) => ( element ) => {
        function onKeyDown( event ) {
-               const { keyCode } = event;
+               const { keyCode, ctrlKey, shiftKey } = event;
 
                if ( event.defaultPrevented ) {
                        return;
@@ -30,6 +30,11 @@ export default ( props ) => ( element ) => {
                                return;
                        }
 
+                       // Exclude command+shift+del/delete as they are shortcuts for deleting blocks.
+                       if ( event.ctrlKey && event.shiftKey ) {
+                               return;
+                       }
+
                        if ( onMerge ) {
                                onMerge( ! isReverse );
                        }
```

yogeshbhutkar committed:
KeyboardShortcuts: Change delete shortcut to primaryShift + Backspace

yogeshbhutkar committed:
KeyboardShortcuts: Update delete shortcut to use `shift + Backspace`

yogeshbhutkar commented:
Hi @t-hamano, thanks for suggesting this approach. I've incorporated it in the latest commit. Also, as per [this comment](https://github.com/WordPress/gutenberg/issues/68139#issuecomment-2559627412), I've implemented `shift+backspace` as the standard shortcut for deleting the blocks.

I've updated the PR Title, Description, and Screencasts/ Screenshots used appropriately to reflect the same.

t-hamano reviewed (approved):
LGTM! Everything seems to work correctly on Windows.

- At the beginning of the text:
  - ✅ Press <kbd>backspace</kbd> key:  The blocks are merged (RichText shortcut)
  - ✅ Press <kbd>shift</kbd>+<kbd>backspace</kbd> key: The block is removed (Shortcut implemented in this PR)
  - ✅ Press <kbd>shift</kbd>+<kbd>del</kbd> key: The block is moved to the clipboard (Chrome shortcut)
- At the end of the text:
  - ✅ Press <kbd>del</kbd> key:  The blocks are merged (RichText shortcut)
  - ✅ Press <kbd>shift</kbd>+<kbd>backspace</kbd> key: The block is removed (Shortcut implemented in this PR)
  - ✅ Press <kbd>shift</kbd>+<kbd>del</kbd> key: The block is moved to the clipboard (Chrome shortcut) 

@talldan @jasmussen Just to be sure, could you check if everything works as expected on Mac as well?

jasmussen reviewed (approved):
Confirmed, this works well on a Mac. It seems more ergonomic, intuitive, predictable, and works well across laptop and extended keyboards. Nice work!

yogeshbhutkar commented:
Thank you, @t-hamano and @jasmussen, for taking the time to test the PR ✨!

afercia commented:
Shift + Backspace is listed as the Firefox shortcut for histoyr 'Forward' action on Windows.
It's unclear to me if that's still true. @t-hamano would you e so king to double check, when you have a chance?

https://support.mozilla.org/en-US/kb/keyboard-shortcuts-perform-firefox-tasks-quickly

![Screenshot 2025-01-13 at 13 04 39](https://github.com/user-attachments/assets/33e87456-ab39-4adf-acc4-e41f311327ac)

https://kb.mozillazine.org/Browser.backspace_action
> Pressing [Backspace] will go back a page in the session history and [Shift]+[Backspace] will go forward. (Default in Windows)

To test with Firefox on other platforms, it is possible to edit Firefox's `about:config` and set `browser.backspace_action` to `0`. This way, Backspace and Shift+Backspace will work as 'Back' and 'Forward`. Still to determine if that's the actual default on Windows.

t-hamano commented:
> Shift + Backspace is listed as the Firefox shortcut for history 'Forward' action on Windows.
> It's unclear to me if that's still true.

This feature appears to have been disabled in Firefox version 87:

> To prevent user data loss when filling out forms, we’ve disabled the Backspace key as a navigation shortcut for the back navigation button. To re-enable the Backspace keyboard shortcut, you can change the about:config preference browser.backspace_action to 0. You can also use the recommended Alt + Left arrow (Command + Left arrow on Mac) shortcut instead.

Source: https://www.mozilla.org/en-US/firefox/87.0/releasenotes/

In fact, I checked it using the Firefox browser, and the value of `browser.backspace_action` was `2`.

afercia commented:
> I checked it using the Firefox browser, and the value of browser.backspace_action was 2.

Cool, thanks. The value is `2` on macOS as well. I guess it's OK then.

Czab1 commented:
Wordpress user here, I'm sure this change was made with good intentions, but this shortcut (`Shift + Backspace`) is way too easy to press by accident and I've triggered it multiple times in just 5 minutes of writing a post.

How can I disable this keyboard shortcut? Or perhaps disable all keyboard shortcuts?

t-hamano commented:
@Czab1 Thanks for the report.

> How can I disable this keyboard shortcut? Or perhaps disable all keyboard shortcuts?

There is currently no way to disable a specific shortcut.

Could you please comment in #68139 with details of what operation you were performing that caused this shortcut to be executed unintentionally? We can reopen #68139 to restart the discussion on what shortcut would be more appropriate.

dcalhoun commented:
It appears these changes introduced a regression for iOS devices: https://github.com/WordPress/gutenberg/issues/69121 

Entire blocks are inadvertently removed when pressing <kbd>Delete</kbd> (backspace) key while the cursor is placed as the end of a sentence after a punctuation and space. The issue occurs for me on the current `trunk` branch. Reverting these changes address the issue.  

Notably, applying the changes proposed in https://github.com/WordPress/gutenberg/pull/69074 also (inadvertently?) resolves the issue.

t-hamano commented:
@dcalhoun Thanks for the report.

So, can we merge #69074? My machine is Windows OS, so I can't build and test the mobile app 😅

yogeshbhutkar commented:
Based on the issue reported above (Ref. [#69121](https://github.com/WordPress/gutenberg/issues/69121), the bug was reproduced by the author on Safari (not a native app). I was also able to replicate this behavior in the [Gutenberg Playground](https://playground.wordpress.net/gutenberg.html) for this PR on a physical iPhone 14 Plus. However, it did not occur on simulators, as noted in the original issue:

> It also only occurs on a physical iOS device, I was unable to reproduce it using iOS simulators.

I believe the issue occurs when using the iPhone's keyboard, and it can't be reproduced on simulators.

On the brighter side, the above issue **_does not occur_** in the new PR. (Ref. https://github.com/WordPress/gutenberg/pull/69074)

| Old PR (#68164) | New PR (#69074) |
| - | - |
|![old-pr](https://github.com/user-attachments/assets/2c709681-e2f8-4a64-9b1a-d1cf051dc26c)|![new-pr-1](https://github.com/user-attachments/assets/0e3eef05-71cb-4d16-8d49-c6ee1e2f38f8)|

t-hamano commented:
Thanks for testing.

I mistakenly thought the issue was with the mobile app and not with iOS devices.

Either way, #69074 does solve the issue, so let's merge it.
